ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -292,32 +292,37 @@ public ReplicationManager(ReplicationConfig replicationConfig, ClusterMapConfig
-            replicationMetrics.createRemoteReplicaErrorMetrics(remoteReplicaInfo);
-            remoteReplicas.add(remoteReplicaInfo);
-            updateReplicasToReplicate(remoteReplica.getDataNodeId().getDatacenterName(), remoteReplicaInfo);
-          }
-          PartitionInfo partitionInfo = new PartitionInfo(remoteReplicas, replicaId.getPartitionId(),
-              storeManager.getStore(replicaId.getPartitionId()), replicaId);
-          partitionsToReplicate.put(replicaId.getPartitionId(), partitionInfo);
-          List<PartitionInfo> partitionInfos = partitionGroupedByMountPath.get(replicaId.getMountPath());
-          if (partitionInfos == null) {
-            partitionInfos = new ArrayList<PartitionInfo>();
+        PartitionId partition = replicaId.getPartitionId();
+        Store store = storeManager.getStore(partition);
+        if (store != null) {
+          List<ReplicaId> peerReplicas = replicaId.getPeerReplicaIds();
+          if (peerReplicas != null) {
+            List<RemoteReplicaInfo> remoteReplicas = new ArrayList<RemoteReplicaInfo>(peerReplicas.size());
+            for (ReplicaId remoteReplica : peerReplicas) {
+              // We need to ensure that a replica token gets persisted only after the corresponding data in the
+              // store gets flushed to disk. We use the store flush interval multiplied by a constant factor
+              // to determine the token flush interval
+              RemoteReplicaInfo remoteReplicaInfo =
+                  new RemoteReplicaInfo(remoteReplica, replicaId, store, factory.getNewFindToken(),
+                      storeConfig.storeDataFlushIntervalSeconds * SystemTime.MsPerSec * Replication_Delay_Multiplier,
+                      SystemTime.getInstance(), remoteReplica.getDataNodeId().getPortToConnectTo());
+              replicationMetrics.addRemoteReplicaToLagMetrics(remoteReplicaInfo);
+              replicationMetrics.createRemoteReplicaErrorMetrics(remoteReplicaInfo);
+              remoteReplicas.add(remoteReplicaInfo);
+              updateReplicasToReplicate(remoteReplica.getDataNodeId().getDatacenterName(), remoteReplicaInfo);
+            }
+            PartitionInfo partitionInfo = new PartitionInfo(remoteReplicas, partition, store, replicaId);
+            partitionsToReplicate.put(partition, partitionInfo);
+            List<PartitionInfo> partitionInfos = partitionGroupedByMountPath.get(replicaId.getMountPath());
+            if (partitionInfos == null) {
+              partitionInfos = new ArrayList<>();
+            }
+            partitionInfos.add(partitionInfo);
+            partitionGroupedByMountPath.put(replicaId.getMountPath(), partitionInfos);
           }
-          partitionInfos.add(partitionInfo);
-          partitionGroupedByMountPath.put(replicaId.getMountPath(), partitionInfos);
+        } else {
+          logger
+              .error("Not replicating to partition " + partition + " because an initialized store could not be found");
```

Comment:
I don't feel like we do because we already have metrics that tell us when stores fail to startup. At this point, we can inspect the logs on a machine to see specifics.